app/Filament/Widgets/Dashboard/TrendExpense.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace App\Filament\Widgets\Dashboard;
+
+use App\Concerns\HasFilterPeriod;
+use App\Enums\Period;
+use App\Models\ExpenseBudget;
+use Carbon\Carbon;
+use Filament\Widgets\Concerns\InteractsWithPageFilters;
+use Flowframe\Trend\Trend;
+use Leandrocfe\FilamentApexCharts\Widgets\ApexChartWidget;
+
+class TrendExpense extends ApexChartWidget
+{
+    use HasFilterPeriod, InteractsWithPageFilters;
+
+    protected static ?string $chartId = 'trendExpense';
+
+    protected function getHeading(): ?string
+    {
+        return __('budgetly::widgets.dashboard.trend_expense');
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    protected function getOptions(): array
+    {
+        /** @var array<int, mixed> $period */
+        $period = $this->getFilterPeriod();
+
+        $filter = $this->filters['period'];
+
+        /**
+         * @var Carbon $startDate
+         * @var Carbon $endDate
+         */
+        [$startDate, $endDate] = $period;
+
+        $trend = Trend::query(ExpenseBudget::query()->whereBelongsToUser(auth()->user()))
+            ->between($startDate, $endDate);
+
+        if ($filter === Period::Today->value) {
+            $trend->perHour();
+        } elseif ($filter === Period::Yesterday->value) {
+            $trend->perDay();
+        } elseif ($filter === Period::LastSevenDays->value) {
+            $trend->perDay();
+        } elseif ($filter === Period::LastMonth->value) {
+            $trend->perWeek();
+        } elseif ($filter === Period::ThisMonth->value) {
+            $trend->perWeek();
+        } elseif ($filter === Period::MonthToDate->value) {
+            $trend->perWeek();
+        } elseif ($filter === Period::YearToDate->value) {
+            $trend->perMonth();
+        } elseif ($filter === Period::Custom->value) {
+            $diffDays = intval($startDate->diffInDays($endDate));
+
+            if ($diffDays > 1) {
+                $trend->perDay();
+            } elseif ($diffDays >= 30) {
+                $trend->perWeek();
+            } elseif ($diffDays > 90) {
+                $trend->perMonth();
+            } else {
+                $trend->perHour();
+            }
+        } else {
+            $trend->perHour();
+        }
```

Comment:
i think to make new handler for this, this is not readable we can make separate class called TrendManager it might something like below 

```php
class TrendManager
{
    private const PERIOD_INTERVALS = [
        Period::Today->value => 'perHour',
        Period::Yesterday->value => 'perDay',
        Period::LastSevenDays->value => 'perDay',
        Period::LastMonth->value => 'perWeek',
        Period::ThisMonth->value => 'perWeek',
        Period::MonthToDate->value => 'perWeek',
        Period::YearToDate->value => 'perMonth',
    ];

    public function setTrendInterval(string $filter, Trend $trend, ?Carbon $startDate = null, ?Carbon $endDate = null): void
    {
        if ($filter === Period::Custom->value) {
            $this->handleCustomPeriod($trend, $startDate, $endDate);

            return;
        }

        $method = self::PERIOD_INTERVALS[$filter] ?? 'perHour';

        $trend->$method();
    }

    private function handleCustomPeriod(Trend $trend, ?Carbon $startDate, ?Carbon $endDate): void
    {
        if (!$startDate || !$endDate) {
            $trend->perHour();

            return;
        }

        $diffDays = (int) $startDate->diffInDays($endDate);

        $interval = match(true) {
            $diffDays > 90 => 'perMonth',
            $diffDays >= 30 => 'perWeek',
            $diffDays > 1 => 'perDay',
            default => 'perHour'
        };

        $trend->$interval();
    }
}
``` 

and we can use it like below 
```php

$trend = Trend::query(ExpenseBudget::query()->whereBelongsToUser(auth()->user()))
            ->between($startDate, $endDate);

(new TrendManager)->setTrendInterval($filter, $trend, $startDate, $endDate)

```

Reply:
I see, thanks